### DISCUSSION TITLE
KV_cache issues

### DISCUSSION THREAD

**Discussion by tqtifnypmb** [openai/whisper#1117] [Q&A] [answered]
It seems that Whisper has cached the results of the `Key` `Value` layer in `MultiHeadAttention`, so when predicting the next token, we only need to pass the previous token. However, the results of the `Query` layer have not been cached, which should be a bug, but the transcription results of Whisper seem to be correct. Am I misunderstanding something?

**Comment by jongwook** [comment, marked as answer]
This is because query projections are not reused across different positions. Some good references on KV cache:

- https://lilianweng.github.io/posts/2023-01-10-inference-optimization/
- https://kipp.ly/blog/transformer-inference-arithmetic/
- https://www.dipkumar.dev/becoming-the-unbeatable/posts/gpt-kvcache/

**Reply by tqtifnypmb** [reply]
Thanks
